src/main/java/com/arpnetworking/tsdcore/sinks/TcpSinkActor.java

```diff
@@ -0,0 +1,282 @@
+            final String serverAddress,
+            final int serverPort,
+            final int maximumQueueSize,
+            final int exponentialBackoffBase) {
+        return Props.create(TcpSinkActor.class, sink, serverAddress, serverPort, maximumQueueSize, exponentialBackoffBase);
+    }
+
+    /**
+     * Public constructor.
+     *
+     * @param sink Sink that controls request creation and data serialization.
+     * @param serverAddress Server to connect to.
+     * @param serverPort Port to connect to.
+     * @param maximumQueueSize Maximum number of pending requests.
+     * @param exponentialBackoffBase Milliseconds as the base as the connection exponential backoff.
+     */
+    public TcpSinkActor(
+            final TcpSink sink,
+            final String serverAddress,
+            final int serverPort,
+            final int maximumQueueSize,
+            final int exponentialBackoffBase) {
+        _sink = sink;
+        _serverAddress = serverAddress;
+        _serverPort = serverPort;
+        _maximumQueueSize = maximumQueueSize;
+        _exponentialBackoffBase = exponentialBackoffBase;
+        _pendingRequests = new LinkedList<>();
+        connect();
+    }
+
+    /**
+     * Generate a Steno log compatible representation.
+     *
+     * @return Steno log compatible representation.
+     */
+    @LogValue
+    public Object toLogValue() {
+        return LogValueMapFactory.builder(this)
+                .put("sink", _sink)
+                .put("serverAddress", _serverAddress)
+                .put("serverPort", _serverPort)
+                .put("exponentialBackoffBase", _exponentialBackoffBase)
+                .build();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String toString() {
+        return toLogValue().toString();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onReceive(final Object message) throws Throwable {
+        if (message instanceof EmitAggregation) {
+            processEmitAggregation((EmitAggregation) message);
+        } else if (message instanceof Ack) {
+            _waitingForAck = false;
+            dispatchPending();
+        } else if (message instanceof Tcp.Connected) {
+            _client = sender();
+            _client.tell(TcpMessage.register(self(), true, true), self());
+        } else if (message instanceof Tcp.CommandFailed) {
+            final Tcp.CommandFailed failed = (Tcp.CommandFailed) message;
+            if (failed.cmd() instanceof Tcp.Connect) {
+                LOGGER.warn()
+                        .setMessage("Failed to connect")
+                        .addData("serverAddress", _serverAddress)
+                        .addData("serverPort", _serverPort)
+                        .log();
+                final int backoffMillis = (((int) (Math.random()  //randomize
+                        * Math.pow(1.3, Math.min(_connectionAttempt, 20)))) //1.3^x where x = min(attempt, 20)
```

Comment:
I'll pull out a constant, but it's been great so far.  I don't think we should encourage changing it.